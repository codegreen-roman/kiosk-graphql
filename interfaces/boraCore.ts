import { Values } from '@interfaces/salesCore'

export interface TimetablePoint {
  date: string
  port: string
  time: string
  timestamp: Date
  locationName: string
}

export const SailStatus = {
  CANCELLED: 'CANCELLED',
  CHECK_IN: 'CHECK-IN',
  CLOSED: 'CLOSED',
  OPEN: 'OPEN',
  TRAVELLED: 'TRAVELLED',
} as const

export type SailStatusType = Values<typeof SailStatus>

export enum SailDirection {
  FORWARD = 'FORWARD',
  BACKWARD = 'BACKWARD',
}

export interface TimetableRow {
  arrival: TimetablePoint
  departure: TimetablePoint
  createdAt: Date
  dangerousGoods: boolean
  departureDate: string
  externalSystemId: string
  route: string
  routeLeg: string
  sailRefId: number
  salesClosed: boolean
  schedule: number
  status: SailStatusType
}

export enum ValueLevel {
  LOW = 'low',
  MIDDLE = 'mid',
  HIGH = 'high',
  HIGH_UPCOMING = 'highUpcoming',
}

export interface DatesDiff {
  isFirstDateNewer: boolean
  isSecondDateNewer: boolean
  isDifferent: boolean
  days: number
  hours: number
  minutes: number
  seconds: number
}

interface TelephoneOutput {
  intlCode: number
  phoneNumber: number
}

export interface ReservationGuest {
  seqN: number
  firstName: string
  lastName: string
  personalIdentificationNumber: string
  __typename: string
}

export interface ReservationVehicle {
  seqN: number
  lengthInCm: number
  widthInCm: number
  heightInCm: number
  weightInKg: number
  type: string
  licencePlate: string
  handicapped: boolean
  companyRegistrationNumber: string
  __typename: string
}
export const OwnerTypeNames = {
  RESERVATION_VEHICLE: 'ReservationVehicle',
  RESERVATION_GUEST: 'ReservationGuest',
}

type ItemOwner = ReservationGuest | ReservationVehicle

export interface ReservationItem {
  quantity: number
  priceCategory: string
  priceCategorySubType: string
  priceCategoryTranslation: string
  pricePerUnit: number
  price: number
  sailPackageSeqN: number
  type: string
  resident: boolean
  inventoryClass: string
  owners: ItemOwner[]
  ownerSeqNs: number[]
  promotion?: string
  promotionText?: string
  promotionPrice?: number
}

interface SailPackage {
  seqN: number
  code: string
  title: string
  sailRefs: SailRef[]
}

interface SailRef {
  sailRefId: number
  route: string
  routeLeg: string
  routeLegTitle: string
  vessel: string
  vesselTitle: string
  departureDate: string
  departureTime: string
  departureAt: string
  port: string
  departureFrom: string
  items: ReservationItem[]
}

type totalPrice = {
  totalPrice: {
    amount: number
    amountFormatted: string
    amountWithoutTaxes: number
    currency: string
    formatPattern: string
    tax: number
    taxAmount: number
  }
}

export enum BoraReservationStatus {
  TEMP = 'TEMP',
  OFFER = 'OFFER',
  BOOKED = 'BOOKED',
  OK = 'OK',
  CANCELLED = 'CN',
  DELETED = 'DL',
}

export enum VehicleWidth {
  LESS_THAN_200 = 'LESS_THAN_200',
  BETWEEN_200_AND_250 = 'BETWEEN_200_AND_250',
  MORE_THAN_250 = 'MORE_THAN_250',
}
export enum VehicleHeight {
  LESS_THAN_400 = 'LESS_THAN_400',
  MORE_THAN_400 = 'MORE_THAN_400',
}

export const VehicleSizes = [
  {
    code: 'SOR-TRI-SOR',
    widthRange: {
      less: {
        value: 190,
        border: '2',
      },
      between: {
        value: 210,
      },
      more: {
        value: 270,
        border: '2.5',
      },
    },
    heightRange: {
      less: {
        value: 390,
        border: '4',
      },
      more: {
        value: 410,
        border: '4',
      },
    },
  },
  {
    code: 'DEFAULT',
    widthRange: {
      less: {
        value: 190,
        border: '2',
      },
      between: {
        value: 210,
      },
      more: {
        value: 270,
        border: '2.6',
      },
    },
    heightRange: {
      less: {
        value: 390,
        border: '4',
      },
      more: {
        value: 410,
        border: '4',
      },
    },
  },
]

export const VehicleLength = {
  MIN: 5,
  MAX: 30,
} as const
export const VehicleWeight = {
  MIN: 2,
  MAX: 70,
} as const

export enum VehicleParameter {
  HEADER = '_HEADER',
  WIDTH = '_WIDTH',
  HEIGHT = '_HEIGHT',
}
export enum PortsToCheckDate {
  SVI = 'SVI',
  ROH = 'ROH',
}

export const companyRegNumberTriggerLength = 8
export const manualIdTriggerLength = 11

export const VesselPrefix = {
  DEFAULT: 'P/L ',
} as const

export const CustomDestination = {
  MUN_MAN: {
    code: 'MUN-MAN',
    map: 'map-man-mun',
  },
  MAN_MUN: {
    code: 'MAN-MUN',
    map: 'map-man-mun',
  },
  MUN_KIH: {
    code: 'MUN-KIH',
    map: 'map-mun-kih',
  },
} as const

export interface ReservationOwnerOutput {
  firstName: string
  lastName: string
  email: string
  phones: TelephoneOutput[]
}
export interface ReservationLoad {
  reservationId: number
  status: BoraReservationStatus
  reservationOwner: ReservationOwnerOutput
  sailPackages: SailPackage[]
  paymentInfo: totalPrice
}

export interface PaymentTerminalOperation {
  operationId: string
}

export type GeneralCodeAndTitle = {
  code: string
  title?: string
  value?: never
}

export type RouteLegPart = {
  location: string
} & GeneralCodeAndTitle

export type RouteLeg = {
  parts: RouteLegPart[]
} & GeneralCodeAndTitle

export type RouteAttribute = GeneralCodeAndTitle

export type Route = {
  legs: RouteLeg[]
  attributes?: RouteAttribute[]
} & GeneralCodeAndTitle

export type KioskSailPackage = { route: Route } & GeneralCodeAndTitle

export type KioskInfo = {
  kioskId: number
  enabled: boolean
  port: string
  sailPackages: KioskSailPackage[]
}

export type KioskInfoData = {
  kiosk: KioskInfo
}
export interface CreditCompanyMember {
  cardNumber: string
  personalIdentificationNumber: string
  customerId: string
  creditCompany: {
    id: string
    name: string
    registrationNumber: string
  }
}
