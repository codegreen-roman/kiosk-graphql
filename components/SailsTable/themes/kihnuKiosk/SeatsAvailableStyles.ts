import styled from 'styled-components'
import { getFonts, getValueLevelColors } from '@themeSettings'
import { ValueLevel } from '@interfaces/boraCore'

const fonts = getFonts()
const valueLevelColors = getValueLevelColors()

interface NumProps {
  level: ValueLevel
}

interface ScaleProps {
  level: ValueLevel
  ratio: number
}

const levelColors = {
  [ValueLevel.LOW]: valueLevelColors.low,
  [ValueLevel.MIDDLE]: valueLevelColors.mid,
  [ValueLevel.HIGH]: valueLevelColors.high,
}

function calculateScaleWidth({ ratio }: ScaleProps): string {
  if (ratio <= 0) return '5px'
  if (ratio >= 1) return '100%'
  return `${100 * ratio}%`
}

export const Root = styled.div`
  display: block;
`

export const NumHolder = styled.div`
  display: flex;
  align-items: flex-start;
`

export const Num = styled.span`
  font: normal 28px/38px ${fonts.secondary};
  color: ${({ level }: NumProps): string => levelColors[level]};
`

export const Local = styled.span`
  font: normal 19px/26px ${fonts.secondary};
  color: ${({ level }: NumProps): string => levelColors[level]};
`

export const Scale = styled.div`
  position: relative;
  width: 35px;
  height: 6px;
  background: ${valueLevelColors.disabled};

  &:after {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: ${calculateScaleWidth};
    height: 100%;
    background: ${({ level }: ScaleProps): string => levelColors[level]};
  }
`
