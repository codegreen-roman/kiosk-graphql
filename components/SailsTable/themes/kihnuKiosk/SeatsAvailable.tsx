import React, { FC } from 'react'
import { Root, Num, NumHolder, Local, Scale } from './SeatsAvailableStyles'
import { getLowAvailableTicketsLevel, getLowReserveTicketsLevel } from '@themeSettings'
import { ValueLevel } from '@interfaces/boraCore'
import { ReservesMax } from '@const/common'

const lowAvailableTicketsLevel = getLowAvailableTicketsLevel()
const lowReserveTicketsLevel = getLowReserveTicketsLevel()

export interface SeatsAvailableProps {
  current: number
  max: number
  local?: number
  reserve?: number
}

interface ComputingLevelColorProps {
  ratio: number
  isReserve: boolean
}
const computeLevelColorForRatio = ({ ratio, isReserve }: ComputingLevelColorProps): ValueLevel => {
  if (ratio === 0 || isNaN(ratio)) {
    return ValueLevel.LOW
  }
  if ((!isReserve && ratio < lowAvailableTicketsLevel) || (isReserve && ratio < lowReserveTicketsLevel)) {
    return ValueLevel.MIDDLE
  }
  return ValueLevel.HIGH
}

const SeatsAvailable: FC<SeatsAvailableProps> = ({ current, max, local, reserve }) => {
  const currentValueFormatted = current === 0 ? `${current}*` : current
  const currentReserveFormatted = reserve === 0 ? `${reserve}**` : reserve

  const currentRatio = current / max
  const currentRatioReserves = reserve / ReservesMax

  const level = computeLevelColorForRatio({ ratio: currentRatio, isReserve: false })
  const levelForReserves = computeLevelColorForRatio({ ratio: currentRatioReserves, isReserve: true })

  return (
    <Root>
      <NumHolder>
        <Num level={level}>{currentValueFormatted}</Num>
        <Local display-if={local} level={level}>
          ({local})
        </Local>
        <Local display-if={reserve >= 0} level={levelForReserves}>
          ({currentReserveFormatted})
        </Local>
      </NumHolder>

      <Scale level={level} ratio={currentRatio} />
    </Root>
  )
}

export default SeatsAvailable
