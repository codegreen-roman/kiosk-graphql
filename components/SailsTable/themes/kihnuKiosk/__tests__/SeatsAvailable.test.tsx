import React, { ReactNode } from 'react'
import SeatsAvailable, { SeatsAvailableProps } from '../SeatsAvailable'
import { Num, Scale } from '../SeatsAvailableStyles'
import { shallow } from 'enzyme'
import { ValueLevel } from '@interfaces/boraCore'
import { getLowAvailableTicketsLevel } from '@themeSettings'

const lowAvailableTicketsLevel = getLowAvailableTicketsLevel()

describe('SeatsAvailable', () => {
  const defaultProps: SeatsAvailableProps = {
    current: 10,
    max: 20,
    local: 5,
  }
  let component

  function render(props?): ReactNode {
    const updatedProps: SeatsAvailableProps = {
      ...defaultProps,
      ...props,
    }
    return shallow(<SeatsAvailable {...updatedProps} />)
  }

  describe('when default props passed', () => {
    beforeAll(() => {
      component = render()
    })

    it('matches snapshot', () => {
      expect(component).toMatchSnapshot()
    })

    it('passes right ratio to scale component', () => {
      expect(component.find(Scale)).toHaveProp('ratio', defaultProps.current / defaultProps.max)
    })
  })

  describe('when current value is 0', () => {
    beforeAll(() => {
      component = render({ current: 0 })
    })

    it('passes LOW level prop to sub components', () => {
      expect(component.find(Num)).toHaveProp('level', ValueLevel.LOW)
      expect(component.find(Scale)).toHaveProp('level', ValueLevel.LOW)
    })
  })

  describe('when current value is lower then lowAvailableTicketsLevel theme setting', () => {
    beforeAll(() => {
      component = render({ current: defaultProps.max * lowAvailableTicketsLevel - 1 })
    })

    it('passes MIDDLE level prop to sub components', () => {
      expect(component.find(Num)).toHaveProp('level', ValueLevel.MIDDLE)
      expect(component.find(Scale)).toHaveProp('level', ValueLevel.MIDDLE)
    })
  })

  describe('when current value is lower then lowAvailableTicketsLevel theme setting', () => {
    beforeAll(() => {
      component = render({ current: defaultProps.max * lowAvailableTicketsLevel + 1 })
    })

    it('passes HIGH level prop to sub components', () => {
      expect(component.find(Num)).toHaveProp('level', ValueLevel.HIGH)
      expect(component.find(Scale)).toHaveProp('level', ValueLevel.HIGH)
    })
  })
})
