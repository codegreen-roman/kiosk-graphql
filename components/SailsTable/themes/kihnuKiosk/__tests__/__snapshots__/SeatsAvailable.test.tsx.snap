// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeatsAvailable when default props passed matches snapshot 1`] = `
<SeatsAvailableStyles__Root>
  <SeatsAvailableStyles__NumHolder>
    <SeatsAvailableStyles__Num
      level="high"
    >
      10
    </SeatsAvailableStyles__Num>
    <SeatsAvailableStyles__Local
      level="high"
    >
      (
      5
      )
    </SeatsAvailableStyles__Local>
  </SeatsAvailableStyles__NumHolder>
  <SeatsAvailableStyles__Scale
    level="high"
    ratio={0.5}
  />
</SeatsAvailableStyles__Root>
`;
